File: src-tauri/src/commands/expiry_lots.rs
//! Tauri commands for expiry lot management: create, update, archive,
//! list, get, and partial resolution.

use tauri::State;

use crate::dto::expiry_lots::{
    ArchiveLotInput, ExpiryLotCreate, ExpiryLotDistributedCreate, ExpiryLotDistributedCreateResult,
    ExpiryLotResolve, ExpiryLotResolveResult, ExpiryLotResponse, ExpiryLotUpdate,
    LotResolutionEventResponse,
};
use crate::error::{AppError, CommandError};
use crate::pdf::locale::Locale;
use crate::services::expiry_lots as service;
use crate::services::user_messages::{
    localize_business_rule, localize_duplicate_field, localize_internal, localize_not_found,
    localize_validation,
};
use crate::state::AppState;

/// Resolves an optional BCP-47 locale tag into a [`Locale`], falling back to
/// English when the frontend does not supply one. Used at every expiry-lot
/// command boundary that accepts an optional locale so dynamic BusinessRule
/// messages (status / quantity / unit) and the adjacent resolve-quantity
/// Validation reach the UI in the active locale while English stays the
/// safe default for legacy callers that omit the argument.
fn resolve_locale(locale: Option<String>) -> Locale {
    locale.as_deref().map(Locale::parse).unwrap_or(Locale::En)
}

/// Returns all active expiry lots across all stores, ordered by expiry date.
#[tauri::command]
pub async fn list_expiry_lots(
    state: State<'_, AppState>,
) -> Result<Vec<ExpiryLotResponse>, CommandError> {
    let pool = state.pool().await;
    service::list_all_expiry_lots(&pool)
        .await
        .map_err(AppError::into)
}

/// Returns all active expiry lots for a given store.
#[tauri::command]
pub async fn list_expiry_lots_by_store(
    state: State<'_, AppState>,
    store_id: String,
) -> Result<Vec<ExpiryLotResponse>, CommandError> {
    let pool = state.pool().await;
    service::list_expiry_lots_by_store(&pool, store_id)
        .await
        .map_err(AppError::into)
}

/// Returns all active expiry lots for a given product, ordered by expiry date.
#[tauri::command]
pub async fn list_expiry_lots_by_product(
    state: State<'_, AppState>,
    product_id: String,
) -> Result<Vec<ExpiryLotResponse>, CommandError> {
    let pool = state.pool().await;
    service::list_expiry_lots_by_product(&pool, product_id)
        .await
        .map_err(AppError::into)
}

/// Returns a single expiry lot by id.
#[tauri::command]
pub async fn get_expiry_lot(
    state: State<'_, AppState>,
    id: String,
) -> Result<ExpiryLotResponse, CommandError> {
    let pool = state.pool().await;
    service::get_expiry_lot(&pool, id)
        .await
        .map_err(AppError::into)
}

/// Creates a new expiry lot. Pre-fills unit and alert-days from the product
/// defaults when those fields are omitted (None). Fails if no active store
/// exists.
///
/// `locale` (BCP-47 tag) is forwarded to `localize_validation` and
/// `localize_business_rule` so the `LocationRequired` validation surfaced
/// by the service and the constant store-required BusinessRule reach the UI
/// in the active locale. Unknown tags fall back to English via
/// `Locale::parse`.
///
/// The argument is required (not `Option<String>`) to preserve the existing
/// frontend contract: `createExpiryLot` already passes `locale` from the
/// wrapper. Keeping it required avoids an IPC signature change for active
/// callers while still chaining both helpers at the boundary.
#[tauri::command]
pub async fn create_expiry_lot(
    state: State<'_, AppState>,
    input: ExpiryLotCreate,
    locale: String,
) -> Result<ExpiryLotResponse, CommandError> {
    let pool = state.pool().await;
    let loc = resolve_locale(Some(locale));
    service::create_expiry_lot(&pool, input)
        .await
        .map_err(|e| localize_validation(e, loc))
        .map_err(|e| localize_business_rule(e, loc))
        .map_err(|e| localize_not_found(e, loc))
        .map_err(|e| localize_duplicate_field(e, loc))
        .map_err(|e| localize_internal(e, loc))
        .map_err(AppError::into)
}

/// Updates an existing active expiry lot.
///
/// `locale` (BCP-47 tag, optional) is forwarded to `localize_validation`
/// and `localize_business_rule` so the dynamic BusinessRule messages
/// (`Cannot update lot: status is ...` and the metadata-only quantity
/// guard) reach the UI in the active locale. Omitting the argument keeps
/// English, preserving backwards compatibility with callers that don't yet
/// pass a locale.
#[tauri::command]
pub async fn update_expiry_lot(
    state: State<'_, AppState>,
    input: ExpiryLotUpdate,
    locale: Option<String>,
) -> Result<ExpiryLotResponse, CommandError> {
    let pool = state.pool().await;
    let loc = resolve_locale(locale);
    service::update_expiry_lot(&pool, input)
        .await
        .map_err(|e| localize_validation(e, loc))
        .map_err(|e| localize_business_rule(e, loc))
        .map_err(|e| localize_not_found(e, loc))
        .map_err(|e| localize_duplicate_field(e, loc))
        .map_err(|e| localize_internal(e, loc))
        .map_err(AppError::into)
}

/// Creates one expiry lot whose initial quantity is distributed across
/// multiple active locations in the same store, atomically.
///
/// The backend emits one `entry:initial` movement for the full lot
/// total to the first allocation's location, then one `transfer` per
/// remaining allocation, all in a single DB transaction. Either every
/// write commits or every write rolls back.
///
/// `locale` (BCP-47 tag, optional) is forwarded through the same
/// `localize_*` chain as `create_expiry_lot` so every user-visible
/// Validation / BusinessRule message (empty distribution, duplicate
/// location, total mismatch, non-store or inactive location,
/// fractional-quantity rejection for integer-unit products) reaches
/// the UI in the active locale. Omitting the argument keeps English.
#[tauri::command]
pub async fn create_expiry_lot_distributed(
    state: State<'_, AppState>,
    input: ExpiryLotDistributedCreate,
    locale: Option<String>,
) -> Result<ExpiryLotDistributedCreateResult, CommandError> {
    let pool = state.pool().await;
    let loc = resolve_locale(locale);
    service::create_expiry_lot_distributed(&pool, input)
        .await
        .map_err(|e| localize_validation(e, loc))
        .map_err(|e| localize_business_rule(e, loc))
        .map_err(|e| localize_not_found(e, loc))
        .map_err(|e| localize_duplicate_field(e, loc))
        .map_err(|e| localize_internal(e, loc))
        .map_err(AppError::into)
}

/// Soft-archives an expiry lot (status = 'archived') with a required
/// justification. Persists the archive reason + notes in `lot_movements`
/// as an `exit:other` marker in the same transaction. Archived lots are
/// excluded from active lists and dashboard queries.
///
/// `locale` (BCP-47 tag, optional) is forwarded to `localize_validation`
/// and `localize_business_rule` so the dynamic BusinessRule messages
/// (`Cannot archive lot: status is already ...` and the race-recovered
/// `Lot is no longer active ...`) reach the UI in the active locale.
/// Omitting the argument keeps English.
#[tauri::command]
pub async fn archive_expiry_lot(
    state: State<'_, AppState>,
    input: ArchiveLotInput,
    locale: Option<String>,
) -> Result<(), CommandError> {
    let pool = state.pool().await;
    let loc = resolve_locale(locale);
    service::archive_expiry_lot(&pool, input)
        .await
        .map_err(|e| localize_validation(e, loc))
        .map_err(|e| localize_business_rule(e, loc))
        .map_err(|e| localize_not_found(e, loc))
        .map_err(|e| localize_duplicate_field(e, loc))
        .map_err(|e| localize_internal(e, loc))
        .map_err(AppError::into)
}

/// Resolves (consumes, discards, or transfers) a quantity from an expiry lot.
/// Records a resolution event and marks the lot as fully resolved when
/// remaining quantity reaches zero.
///
/// `locale` (BCP-47 tag, optional) is forwarded to `localize_validation`
/// and `localize_business_rule` so both surfaces — the
/// `Cannot resolve lot: lot is already ...` BusinessRule and the
/// `Cannot resolve {qty} {unit}: only {n} {unit} remain` Validation —
/// reach the UI in the active locale. Omitting the argument keeps
/// English.
#[tauri::command]
pub async fn resolve_expiry_lot(
    state: State<'_, AppState>,
    input: ExpiryLotResolve,
    locale: Option<String>,
) -> Result<ExpiryLotResolveResult, CommandError> {
    let pool = state.pool().await;
    let loc = resolve_locale(locale);
    service::resolve_expiry_lot(&pool, input)
        .await
        .map_err(|e| localize_validation(e, loc))
        .map_err(|e| localize_business_rule(e, loc))
        .map_err(|e| localize_not_found(e, loc))
        .map_err(|e| localize_internal(e, loc))
        .map_err(AppError::into)
}

/// Returns all resolution events for a given expiry lot, newest first.
#[tauri::command]
pub async fn list_lot_resolution_events(
    state: State<'_, AppState>,
    lot_id: String,
) -> Result<Vec<LotResolutionEventResponse>, CommandError> {
    let pool = state.pool().await;
    service::list_resolution_events(&pool, lot_id)
        .await
        .map_err(AppError::into)
}

File: src/lib/expiry_lots.ts
/**
 * TypeScript API wrapper for Caduxo Tauri commands (Slice 5a — expiry lots).
 * Thin wrappers around the Rust command layer; no business logic here.
 */

import { invoke } from "@tauri-apps/api/core";
import type { UnitKind } from "./products.js";
import {
   DEFAULT_LOCALE,
   locale as activeLocale,
   type SupportedLocale,
} from "../i18n/locale.svelte.js";

// ─── DTOs ────────────────────────────────────────────────────────────────────

export interface ExpiryLotCreate {
  product_id: string;
  store_id: string;
  location_id?: string | null;
  quantity: number;
  unit?: string | null;
  expiry_date: string;
  alert_days_before?: number | null;
  batch_code?: string | null;
  notes?: string | null;
}

export interface ExpiryLotUpdate {
  id: string;
  location_id?: string | null;
  quantity: number;
  unit: string;
  expiry_date: string;
  alert_days_before: number;
  batch_code?: string | null;
  notes?: string | null;
}

export interface ExpiryLotResolve {
  lot_id: string;
  quantity: number;
  resolution: string;
  notes?: string | null;
}

/**
 * Allow-listed archive reason codes. Mirrors the backend
 * `VALID_ARCHIVE_REASONS` constant in `src-tauri/src/services/expiry_lots.rs`.
 * Keep both in sync when adding new reasons.
 */
export interface ArchiveReasonOption {
  value: string;
  label: string;
}

export const ARCHIVE_REASONS: readonly ArchiveReasonOption[] = [
  { value: "expired_unsold", label: "Expired (unsold)" },
  { value: "damaged", label: "Damaged" },
  { value: "returned_to_supplier", label: "Returned to supplier" },
  { value: "recall", label: "Manufacturer recall" },
  { value: "lost", label: "Lost / unaccounted" },
  { value: "internal_use", label: "Internal use" },
  { value: "administrative", label: "Administrative cleanup" },
  { value: "other", label: "Other" },
] as const;

/** Minimum trimmed length required for archive `notes` (server-side rule). */
export const ARCHIVE_NOTES_MIN_CHARS = 5;
/** Maximum trimmed length allowed for archive `notes` (server-side rule). */
export const ARCHIVE_NOTES_MAX_CHARS = 1000;

/** Input for archiving an active expiry lot with a required justification. */
export interface ArchiveLotInput {
  id: string;
  reason: string;
  notes: string;
}

/**
 * One (location_id, quantity) pair within a distributed lot creation.
 *
 * Allocations are applied atomically by `createExpiryLotDistributed`.
 * The backend rejects empty arrays, duplicate `location_id`s,
 * quantities `<= 0`, fractional quantities for integer-unit products,
 * and locations that don't belong to the lot's `store_id` or are
 * inactive. The frontend should mirror these rules in the form layer
 * so the user gets inline feedback before the IPC round-trip.
 */
export interface ExpiryLotAllocationInput {
  location_id: string;
  quantity: number;
}

/**
 * Input for creating one expiry lot distributed across multiple active
 * locations in the same store. Mirrors
 * `dto::expiry_lots::ExpiryLotDistributedCreate` in the Rust service.
 *
 * `total_quantity` is the explicit lot total and must equal the sum of
 * allocation quantities (within a tiny float tolerance); a mismatch is
 * rejected at the IPC boundary with `DistributionTotalMismatch`.
 */
export interface ExpiryLotDistributedCreate {
  product_id: string;
  store_id: string;
  total_quantity: number;
  allocations: ExpiryLotAllocationInput[];
  unit?: string | null;
  expiry_date: string;
  alert_days_before?: number | null;
  batch_code?: string | null;
  notes?: string | null;
}

/**
 * Result of a distributed lot creation. `lot` is the persisted expiry
 * lot; `anchor_location_id` is the location that received the single
 * `entry:initial` movement (i.e. the first allocation); the per-location
 * ledger can be re-read through `getLotLocationBalances`.
 */
export interface ExpiryLotDistributedCreateResult {
  lot: ExpiryLotResponse;
  anchor_location_id: string;
  allocation_count: number;
}

export interface ExpiryLotResponse {
  id: string;
  product_id: string;
  store_id: string;
  location_id: string | null;
  /**
   * Human-readable location name projected from `store_locations.name` via
   * the backend `LEFT JOIN` (see `odd/tasks/lot-location-name.md`).
   * `null` when the lot has no `location_id` or the row is missing;
   * inactive locations are still included so historic lots stay
   * interpretable. UI display helpers consume this to render the name
   * instead of the raw `location_id`; treat `location_id` as the
   * authoritative relationship key and `location_name` as a presentational
   * projection that may be absent on legacy rows.
   */
  location_name: string | null;
  quantity: number;
  unit: string;
  expiry_date: string;
  alert_days_before: number;
  batch_code: string | null;
  status: string;
  resolution: string | null;
  resolved_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  /**
   * Unit kind resolved from the product's catalog link.
   * `null` for legacy / uncatalogued products — treated as decimal by the UI.
   * Drives dynamic min/step/validation rules in movement modals.
   */
  unit_type: UnitKind | null;
}

export interface LotResolutionEventResponse {
  id: string;
  expiry_lot_id: string;
  quantity: number;
  resolution: string;
  notes: string | null;
  created_at: string;
}

export interface ExpiryLotResolveResult {
  lot_id: string;
  resolved_quantity: number;
  remaining_quantity: number;
  is_fully_resolved: boolean;
  resolution_event_id: string;
}

// ─── Commands ─────────────────────────────────────────────────────────────────

/** Returns all active expiry lots across all stores, ordered by expiry date. */
export async function listExpiryLots(): Promise<ExpiryLotResponse[]> {
  return invoke<ExpiryLotResponse[]>("list_expiry_lots");
}

/** Returns all active expiry lots for a given store. */
export async function listExpiryLotsByStore(
  storeId: string,
): Promise<ExpiryLotResponse[]> {
  return invoke<ExpiryLotResponse[]>("list_expiry_lots_by_store", {
    storeId,
  });
}

/** Returns all active expiry lots for a given product, ordered by expiry date. */
export async function listExpiryLotsByProduct(
  productId: string,
): Promise<ExpiryLotResponse[]> {
  return invoke<ExpiryLotResponse[]>("list_expiry_lots_by_product", {
    productId,
  });
}

/** Returns a single expiry lot by id. */
export async function getExpiryLot(id: string): Promise<ExpiryLotResponse> {
  return invoke<ExpiryLotResponse>("get_expiry_lot", { id });
}

/**
 * Returns the active UI locale, used as the implicit default for the
 * `locale` parameter accepted by every backend-call wrapper in this file.
 *
 * Callers that already hold a `SupportedLocale` can pass it explicitly to
 * avoid the rune read at call time; otherwise the wrapper reads
 * `activeLocale.current` and falls back to `DEFAULT_LOCALE` while the rune
 * is still uninitialised.
 */
function resolveLocale(locale?: SupportedLocale): SupportedLocale {
   if (locale) return locale;
   const current = activeLocale?.current;
   return (current ?? DEFAULT_LOCALE) as SupportedLocale;
}

/**
 * Creates a new expiry lot.
 *
 * `locale` is forwarded to the Rust command so the backend can localise
 * validation errors (notably `LocationRequired`) before they reach the UI.
 * When omitted, the wrapper uses the active UI locale.
 */
export async function createExpiryLot(
  input: ExpiryLotCreate,
  locale?: SupportedLocale,
): Promise<ExpiryLotResponse> {
  return invoke<ExpiryLotResponse>("create_expiry_lot", {
    input,
    locale: resolveLocale(locale),
  });
}

/**
 * Creates one expiry lot with the initial quantity distributed across
 * multiple active locations in the same store, atomically.
 *
 * The backend writes one `entry:initial` movement for the full lot total
 * to the first allocation's location, then one `transfer` per remaining
 * allocation, all in a single DB transaction. Either every write
 * commits or every write rolls back — there is no partial commit path.
 *
 * `locale` is forwarded to the Rust command so the backend can localise
 * the new distributed-creation Validation / BusinessRule messages
 * (empty distribution, duplicate location, total mismatch, non-store or
 * inactive location, fractional-quantity rejection for integer-unit
 * products) before they reach the UI. When omitted, the wrapper uses
 * the active UI locale.
 */
export async function createExpiryLotDistributed(
  input: ExpiryLotDistributedCreate,
  locale?: SupportedLocale,
): Promise<ExpiryLotDistributedCreateResult> {
  return invoke<ExpiryLotDistributedCreateResult>("create_expiry_lot_distributed", {
    input,
    locale: resolveLocale(locale),
  });
}

/** Updates an existing active expiry lot. */
export async function updateExpiryLot(
  input: ExpiryLotUpdate,
): Promise<ExpiryLotResponse> {
  return invoke<ExpiryLotResponse>("update_expiry_lot", { input });
}

/**
 * Soft-archives an active expiry lot with a required justification.
 *
 * The Rust command persists the archive reason + notes in `lot_movements`
 * as an `exit:other` marker in the same transaction as flipping the lot
 * status to `archived`. Pass `notes` as trimmed text — the backend will
 * also enforce trimming, a minimum of 5 characters, and a cap of 1000.
 */
export async function archiveExpiryLot(input: ArchiveLotInput): Promise<void> {
  return invoke<void>("archive_expiry_lot", { input });
}

/**
 * Resolves (consumes, discards, or transfers) a quantity from a lot.
 * Records a resolution event and marks the lot as fully resolved when
 * remaining quantity reaches zero.
 */
export async function resolveExpiryLot(
  input: ExpiryLotResolve,
): Promise<ExpiryLotResolveResult> {
  return invoke<ExpiryLotResolveResult>("resolve_expiry_lot", { input });
}

/** Returns all resolution events for a given lot, newest first. */
export async function listLotResolutionEvents(
  lotId: string,
): Promise<LotResolutionEventResponse[]> {
  return invoke<LotResolutionEventResponse[]>("list_lot_resolution_events", {
    lotId,
  });
}
